mmap-alloc/src/lib.rs

```diff
@@ -295,6 +295,56 @@ impl MapAlloc {
         map(size, self.perms, self.commit, self.huge_pagesize).and_then(f)
     }
 
+    #[cfg(target_os = "linux")]
+    unsafe fn realloc_helper(&self, old_ptr: *mut u8, old_size: usize, new_size: usize) -> Option<*mut u8> {
+        // Since allocators in Rust are not allowed to return null pointers, but it may be valid for
+        // mremap to return memory starting at null, we have to handle that case. We
+        // do this by checking for null, and if we find that map has returned null, we create a
+        // new mapping, and mremap the data onto that new mapping, and try to map a new empty page
+        // at address 0. Since we leave the page at 0 mapped, future calls should never return null.
+        // Note that this leaks memory since we never unmap that page, but this isn't a big deal -
+        // since we never write to it, it will remain uncommitted and will thus not consume any
+        // physical memory.
+        let f = |ptr: *mut u8| if ptr.is_null() {
+            let new_ptr = map(new_size, self.perms, false, self.huge_pagesize)
+                .expect("Not enough virtual memory to make new mapping");
+            debug_assert!(!new_ptr.is_null(),
+                "we have an open mapping on null, new mapping should never be null");
+
+            let move_result = libc::mremap(
+                ptr::null_mut(),
+                new_size,
+                new_size,
+                libc::MREMAP_MAYMOVE | libc::MREMAP_FIXED,
+                new_ptr
+            );
+            if move_result == libc::MAP_FAILED {
+                panic!("Unable to move mapping: {}", errno());
+            }
+
+            let pagesize = sysconf::page::pagesize();
+            let null_map = libc::mmap(
+                ptr::null_mut(),
+                pagesize,
+                libc::PROT_NONE,
+                libc::MAP_ANONYMOUS | libc::MAP_FIXED | libc::MAP_NORESERVE,
+                -1,
+                0
+            );
+            // ignore errors creating the new map at null
```

Comment:
Clarify that we're ignoring the error because any error implies that somebody else mapped this page in between us unmapping it and attempting to map it again.

Reply:
Even in that case, we shouldn't get an error. One of the nice/scary things about MAP_FIXED, is:

> If the memory region specified by addr and len overlaps pages of
              any existing mapping(s), then the overlapped part of the
              existing mapping(s) will be discarded.

Reply:
Ah, crap. That means that we'll clobber the contents of that page, right? I think we're gonna need to figure out another way to do that because that's a race condition even with our own code (e.g., if the thread that mapped the page was another instance of this function, and we clobber their data before they call `mremap`).

Reply:
That's true. I don't see any way around that without some sort of mutex, or just leaving it unmapped, and hoping someone gets a null back from normal allocate (which can safely leave an empty page mapped)

Reply:
Can we just do the memcpy manually? Store the first byte like I mentioned before, and then memcpy from the existing allocation to the new one. Once that's done, all but the first page of the existing allocation can be safely unmapped.

Reply:
Given that the mutex would only have to be around [these lines](https://github.com/Dr-Emann/allocators-rs/blob/ddb98c5df261be05a8e13a94fd8d3dcc9dcb8a16/mmap-alloc/src/lib.rs#L314-L340), and the chances of contention are basically 0, and it's only in the case where mremap returns null, I lean toward the mutex.

Reply:
Oh, right. no_std makes that harder.

Reply:
no_std, and also the fact that other code besides ours might be responsible for the mapping.

Reply:
If we do the memcpy, what should we do about mprotect?

Reply:
I think we'll have to bite the bullet and mprotect read the source, mprotect write the destination, do the memcpy, and then set the permissions back to what they should be.